addition: api response examples + spring generator can generate response examples (#17610, #16051)

GlobeDaBoarder opened the pull request:
<!-- Enter details of the change here. Include additional tests that have been done, reference to the issue for tracking, etc. -->

# Addition of API response examples to the core generator + Spring's `api.mustache` generates response examples  

This PR closes issues #17610 and #16051

## Reasons for creating this PR

Having response examples results in a more detailed and understandable API documentation, which covers all possible variations of the same response codes. For more details refer to the [issue](https://github.com/OpenAPITools/openapi-generator/issues/17610). This is particularly useful for integrating the front-end with the back-end. Having those corner cases and exact response examples covered in the documentation allows for a clear way of handling things on the front-end side.

## What was done

NOTE: There are 2 commits in this PR: the feature changes in the code itself and the sample updates. Due to the fact that updating samples introduced 208 file changes _(I suppose some PRs merged before opening this one did not run the update samples script_), please refer to the [feature commit](https://github.com/OpenAPITools/openapi-generator/commit/cb514a9c0e811a4b2a79590dd7b56eebb75ade33) to view only the code changes made to the generator, since viewing the whole diff of this PR just introduces too many changes because of sample updates.

### Changes introduced with the feature request:

#### 1. [Changes to the `DefaultCodegen` class.](https://github.com/OpenAPITools/openapi-generator/commit/cb514a9c0e811a4b2a79590dd7b56eebb75ade33#diff-3a138675e8cb40943bcb00feb46edd6f6ee5c5306c29cc19f02c845a299c8658R4920-R4926)

Response examples are retrieved from the [`ApiResponse`](https://github.com/swagger-api/swagger-core/blob/master/modules/swagger-models/src/main/java/io/swagger/v3/oas/models/responses/ApiResponse.java) object and put into [`CodegenResponseObject`](https://github.com/OpenAPITools/openapi-generator/blob/64c763b87ac839e3bd6a4017a743197e2e6ab701/modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenResponse.java#L32) if they are present. If examples are also links ($ref) they are dereferenced/unaliased. This is done with an [`ExampleUtils`](https://github.com/OpenAPITools/openapi-generator/commit/cb514a9c0e811a4b2a79590dd7b56eebb75ade33#diff-fafd68e157413c6679d4e7deb7bc8582f4e176b64ecd4c73ae4e10f457601e3f) file i have created. 

#### 2. Creation of [`ExampleUtils`](https://github.com/OpenAPITools/openapi-generator/commit/cb514a9c0e811a4b2a79590dd7b56eebb75ade33#diff-fafd68e157413c6679d4e7deb7bc8582f4e176b64ecd4c73ae4e10f457601e3f) class 

This class has been heavily inspired by an already existing [`ModelUtils`](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/ModelUtils.java)

The main purpose of this class is:
- retrieve [`Example`](https://github.com/swagger-api/swagger-core/blob/master/modules/swagger-models/src/main/java/io/swagger/v3/oas/models/examples/Example.java) objects from the `ApiReponse` class
- unalias examples (resolve references to actual objects if they are present) 

#### 3. [Adjust Spring's `api.mutache`](https://github.com/OpenAPITools/openapi-generator/commit/cb514a9c0e811a4b2a79590dd7b56eebb75ade33#diff-5cdb43ab5b2469625bc1f47813b6d826768bd48576091fd0ffb7f85c1080c254) file to generate `@ExampleObject` swagger annotations with example content

After adding response examples in the DefaultCodegen, we can utilize that in templates. I primarily use spring generator so added to the template file to be able to generate examples if they are present.

As a result, `@ExamplesObject` swagger annotations will be generated:

![image](https://github.com/OpenAPITools/openapi-generator/assets/74022878/1cd448a3-e0c4-41ee-bfe3-a0a89bd10742)

#### 4. Test case to cover annotation generation 

Also made a little test that recursively checks to see if the `@ExampleObject` annotation is present in the generated Api class. 
The openapi spec resource for the test is this. 


## Result

![image](https://github.com/OpenAPITools/openapi-generator/assets/74022878/f8bfdec3-b112-4c42-a14c-fbfd8edbe737)

## Referencing technical committee

Since this PR includes both a core generator aspect as well as a spring aspect, I hope it is appropriate to mention the spring committee team here:

@cachescrubber, @welshm, @MelleD,  @atextor, @manedev79, @javisst, @borsch, @banlevente, @Zomzog, @martin-mfg 


## Final notes

Please feel free to leave your feedback about the changes and correct me on some things if I am wrong. If this PR ends up getting merged, would also love to see other generators take advantage of response examples : ) 

<!-- Please check the completed items below -->
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ```
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files.
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master.
  These must match the expectations made by your contribution.
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`.
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.1.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

martin-mfg commented:
Thanks for the PR!

Running the spring generator with your changes on your sample input api-response-examples_issue17610.yaml generates output files which don't compile, because:
`src/main/java/org/openapitools/model/Error.java:[9,26] package jakarta.validation does not exist`.
It looks like the corresponding dependency is missing.

I'd suggest adding one or more examples to one or more of the input files from which the [samples](https://github.com/OpenAPITools/openapi-generator/tree/master/samples) are generated. Then the build pipelines of the OpenAPI Generator project will ensure that the generator output is always compilable.

GlobeDaBoarder commented:
Hey, @martin-mfg! Thanks a lot for pointing that out! And thank you for your quick response and feedback. 

I have added 2 sample config files: one with the simple spec file I had before and that you tested out, as well as a classic petstore.yaml spec file which I modified a bit to have response examples.

Btw, when creating a slightly modified petstore sample, I realized that my code didn't take into account the possibility of the presence of XML content media types. My apologies for that oversight. I have adjusted the `ExampleUtils` class and the `api.mustache` a bit to only allow response example generation for `application/json` content types. Well at least for now : ) 

Hopefully, my improvements are the right step forward. Either way, looking forward to any other opinions and recommendations 🙌

GlobeDaBoarder commented:
Oh and btw, I am noticing that one circleci check fails. What's interesting is that it is under `samples/client/petstore/java/okhttp-gson-3.1` which I believe has absolutely nothing to do with my commits and changes. Could that be related to "update samples" command taking some previous commits? Is that a reason for concert in this PR?
